action.yaml

```diff
@@ -1,24 +1,151 @@
 description:
-  GitHub Action that installs and provisions supported tools for workflow steps
-  in self-hosted runners. This relies on the agent having supported tooling
-  installed.
-outputs:
-  go:
-    description:
-      The version of golang that has been installed and is ready for use.
+  "Set up tools e.g Python, Node.js, Go, Java, and Terraform using standard
+  setup actions"
+
+inputs:
+  python:
+    description: "Python version (e.g. 3.11)"
+    required: false
   node:
-    description:
-      The verison of Node.js that has been installed and is ready for use.
+    description: "Node.js version (e.g. 18)"
+    required: false
+  go:
+    description: "Go version (e.g. 1.22)"
+    required: false
   java:
     description:
-      The version of Java that has been installed and is ready for use.
-  python:
-    description:
-      The version of Python that has been installed and is ready for use.
+      "Java version with distribution (e.g. 8.0.442-tem) which is the same as in
+      .sdkmanrc"
+    required: false
   terraform:
-    description:
-      The version of Terraform that has been installed and is ready for use.
+    description: "Terraform version (e.g. 1.6.6)"
+    required: false
+
 runs:
-  using: "node20"
-  main: "dist/main/index.js"
+  using: composite
+  steps:
+    - id: detect-python
+      shell: bash
+      run: |
+        if [ -n "${{ inputs.python }}" ]; then
+          echo "PYTHON_VERSION=${{ inputs.python }}" >> $GITHUB_ENV
+        elif [ -f .python-version ]; then
+          echo "PYTHON_VERSION=$(cat .python-version)" >> $GITHUB_ENV
+        else
+          echo "PYTHON_VERSION=" >> $GITHUB_ENV
+        fi
+
+    - name: Setup Python
+      if: env.PYTHON_VERSION != ''
+      uses: actions/setup-python@v5
+      with:
+        python-version: ${{ env.PYTHON_VERSION }}
+
+    - id: detect-node
+      shell: bash
+      run: |
+        if [ -n "${{ inputs.node }}" ]; then
+          echo "NODE_VERSION=${{ inputs.node }}" >> $GITHUB_ENV
+        elif [ -f .node-version ]; then
+          echo "NODE_VERSION=$(cat .node-version)" >> $GITHUB_ENV
+        elif [ -f .nvmrc ]; then
+          echo "NODE_VERSION=$(cat .nvmrc)" >> $GITHUB_ENV
+        else
+          echo "NODE_VERSION=" >> $GITHUB_ENV
+        fi
+
+    - name: Setup Node.js
+      if: env.NODE_VERSION != ''
+      uses: actions/setup-node@v4
+      with:
+        node-version: ${{ env.NODE_VERSION }}
+
+    - id: detect-go
+      shell: bash
+      run: |
+        if [ -n "${{ inputs.go }}" ]; then
+          echo "GO_VERSION=${{ inputs.go }}" >> $GITHUB_ENV
+        elif [ -f .go-version ]; then
+          echo "GO_VERSION=$(cat .go-version)" >> $GITHUB_ENV
+        else
+          echo "GO_VERSION=" >> $GITHUB_ENV
+        fi
+
+    - name: Setup Go
+      if: env.GO_VERSION != ''
+      uses: actions/setup-go@v4
+      with:
+        go-version: ${{ env.GO_VERSION }}
+
+    - id: detect-java
+      shell: bash
+      run: |
```

Comment:
It might be better to put this snippet in a separate file (maybe even a JS snippet?) so we can write unit tests for it